implement unselectable features

warm-coolguy opened the pull request:
## Summary

- For the GFI, a method has been added to mark features as unselectable. They are not styled differently, but just ignored. If visible/displayed in another way, styling has to remove them there, too.
- For the `extendedMasterportalapiMarkers` in the core, the feature has been doubled; since it is orthogonal to GFI, two configurations are required to get the feature running in this mode, one on the markers, one in the GFI. However, styling has been resolved here, as it is internal to this feature.

## Instructions for local reproduction and review

Use the snowbox to check the behaviour. Use cases have been added.

- GFI (pure): Switch the "Compensation area" layer visible. The feature left of the Flugzeugwerft (left of Finkenwerder) should not be clickable, but the one to its right should be. ![example1](https://github.com/user-attachments/assets/359ba5ac-b966-4716-8df2-d055b63c5a14)
- GFI + `extendedMasterportalapiMarkers`: Switch the "Reports (MML)" layer on. Arbitrarily, features with uneven ID have been deemed unselectable. Check whether the behaviour seems plausible/correct to you.

You may also test GFI on the GeoJSON layer by removing the `extendedMasterportalapiMarkers` and `enableClustering` related code. In that case, all pins are visible, but only pins with even number display a GFI window. As in the first case, styling will have to be added later on the layer itself.

Also, please confirm that nothing in the Meldemichel broke.

dopenguin commented on packages/core/README.md:
```diff
@@ -178,8 +178,9 @@ To figure out the name of the locales to override, inspect the matching plugin i
 | dispatchOnMapSelect | string[]? | If set, the parameters will be spread to dispatchment on map selection. `['target', 'value']` will `dispatch(...['target', 'value'])`. This can be used to open the iconMenu's GFI with `['plugin/iconMenu/openMenuById', 'gfi']`, should the IconMenu exist and the gfi plugin be in it with this id. |
 | hoverStyle | MarkerStyle? | Used as map marker style for hovered features. The default fill color for these markers is `'#7B1045'`. |
+| isSelectable | ((feature: GeoJsonFeature) => boolean)? | If undefined, all features are selectable. If defined, this can be used to sort out features to be unselectable, and such features will be styled different and won't react on click. |
 | selectionStyle | MarkerStyle? | Used as map marker style for selected features. The default fill color for these markers is `'#679100'`. |
-
+| unselectableStyle | MarkerStyle? | Used as a map marker style for unselectable features. Features are unselectable if a given `isSelectable` method returns falsy for a feature. The default fill color for these markers is `'#333333'`. |
```
These changes need a changelog entry

warm-coolguy replied:
https://github.com/Dataport/polar/pull/197/commits/041ab1827791be04535b0aec498e5ab7969a89dd

dopenguin commented on packages/clients/snowbox/src/polar-client.ts:
```diff
@@ -1,8 +1,10 @@
 /* eslint-disable @typescript-eslint/no-non-null-assertion */
 import polarCore from '@polar/core'
 import { changeLanguage } from 'i18next'
+// NOTE bad pattern, but probably fine for a test client
+import { enableClustering } from '../../meldemichel/src/utils/enableClustering'
```
Agreed. This can stay for the snowbox 🎩

dopenguin reviewed (changes requested):
If I configure `featureList`, non-selectable features are still selectable. Thus, they also change their styling if `bindWithCoreHoverSelect` is used.
I would have expected that features that not selectable in the map are not being shown in the `featureList`. Would you agree?

Edit: Meldemichel seems to be working fine as far as I can tell.

🏓 @warm-coolguy

warm-coolguy commented on packages/core/src/vuePlugins/vuex.ts:
```diff
@@ -141,6 +141,12 @@ export const makeStore = () => {
         noop(state.selected)
         return selected
       },
+      selectedCoordinates: (state) => {
+        noop(state.selected)
+        return selected === null
+          ? null
+          : (selected.getGeometry() as Point).getCoordinates()
+      },
```
This addition was required for the GFI to react to the selected feature directly. In the Meldemichel, the reaction is solved differently via a `dispatchOnMapSelect` that opens the GFI as part of the menu, which is not the way for this plugin in general, since the GFI may be not part of the IconMenu, as it is in the Snowbox.

warm-coolguy commented:
> If I configure `featureList`, non-selectable features are still selectable. Thus, they also change their styling if `bindWithCoreHoverSelect` is used. I would have expected that features that not selectable in the map are not being shown in the `featureList`. Would you agree?
> 
> 🏓 @warm-coolguy

Yes, I agree. As discussed before, I'm shoveling the PR over to you and adding you as assignee. Should any question about my implementation so far arise, call me anytime.

dopenguin commented:
> > If I configure `featureList`, non-selectable features are still selectable. Thus, they also change their styling if `bindWithCoreHoverSelect` is used. I would have expected that features that not selectable in the map are not being shown in the `featureList`. Would you agree?
> > 🏓 @warm-coolguy
> 
> Yes, I agree. As discussed before, I'm shoveling the PR over to you and added you as assignee. Should any question about my implementation so far arise, call me anytime.
> 
> 🏓 @dopenguin

I've fixed that behaviour in c38065c41f250ce841c3667d41b134d3cc17c3e8.
If you want to reproduce it, I've attached a [patch](https://github.com/user-attachments/files/17512585/featureList_in_snowbox.patch) with the relevant changes to `@polar/client-snowbox`.

🏓 @warm-coolguy

warm-coolguy commented:
> > > If I configure `featureList`, non-selectable features are still selectable. Thus, they also change their styling if `bindWithCoreHoverSelect` is used. I would have expected that features that not selectable in the map are not being shown in the `featureList`. Would you agree?
> > > 🏓 @warm-coolguy
> > 
> > 
> > Yes, I agree. As discussed before, I'm shoveling the PR over to you and added you as assignee. Should any question about my implementation so far arise, call me anytime.
> > 🏓 @dopenguin
> 
> I've fixed that behaviour in [c38065c](https://github.com/Dataport/polar/commit/c38065c41f250ce841c3667d41b134d3cc17c3e8). If you want to reproduce it, I've attached a [patch](https://github.com/user-attachments/files/17512585/featureList_in_snowbox.patch) with the relevant changes to `@polar/client-snowbox`.
> 
> 🏓 @warm-coolguy

🚀 @dopenguin Checked and just works. Nice.

dopenguin reviewed (approved):
> > > > If I configure `featureList`, non-selectable features are still selectable. Thus, they also change their styling if `bindWithCoreHoverSelect` is used. I would have expected that features that not selectable in the map are not being shown in the `featureList`. Would you agree?
> > > > 🏓 @warm-coolguy
> > > 
> > > 
> > > Yes, I agree. As discussed before, I'm shoveling the PR over to you and added you as assignee. Should any question about my implementation so far arise, call me anytime.
> > > 🏓 @dopenguin
> > 
> > 
> > I've fixed that behaviour in [c38065c](https://github.com/Dataport/polar/commit/c38065c41f250ce841c3667d41b134d3cc17c3e8). If you want to reproduce it, I've attached a [patch](https://github.com/user-attachments/files/17512585/featureList_in_snowbox.patch) with the relevant changes to `@polar/client-snowbox`.
> > 🏓 @warm-coolguy
> 
> 🚀 @dopenguin Checked and just works. Nice.

🚀 @warm-coolguy weeeee
